### PR TITLE
Use References to build Context stores and encode provenance data

### PR DESCRIPTION
**Pull request by sjmiles**
Previously when an Entity was shared from user A to user B, the Entity was copied and it's id was decorated with provenance information, e.g. the source User and source Arc. 

In this PR, shared Entities are presented using share-types that encode a reference to the original Entity and provenance data as proper fields.

For example, `AvatarShare` entities look like:

```
schema AvatarShare extends Share
  Reference<Avatar> ref
  Text fromUser
  Text fromArc
```

Using the references avoids duplication of data in the BackingStore, removes hacky id-encoding, and avoids fetching the contents of `ref` until explicit dereference.  Also, using References.

### CODE REVIEW

**Comment by sjmiles** [comment]
Fixes #2918

File: src/runtime/particle-execution-host.ts [hunk 1]
```diff
@@ -138,7 +138,7 @@ export class ParticleExecutionHost {
         //
         // Without an auditor on the runtime side that inspects what is being fetched from
         // this store, particles with a reference can access any data of that reference's type.
-        this.GetBackingStoreCallback(store, callback, type.collectionOf(), type.toString(), store.id, storageKey);
+        this.GetBackingStoreCallback(store, callback, type.collectionOf(), type.toString(), String(Math.random())/*store.id*/, storageKey);
```
Comment (shaper) [inline src/runtime/particle-execution-host.ts]: Is this ok to merge?

Comment (sjmiles) [reply src/runtime/particle-execution-host.ts]: Good catch, I forgot to follow up on that. I remember discussing it with Myk, but I can't find any reference to it now. :(

Will have to page in the details, make an issue ticket, and put a link here.

Comment (sjmiles) [reply src/runtime/particle-execution-host.ts]: Ok, so I *had* captured the basic problem in #2936.

I tweaked the workaround as Myk suggested and added a note referencing the above ticket.

**Comment by sjmiles** [comment]
Fixes #2728

**Comment by sjmiles** [comment]
Superseded by #3118

**Pull request closed without merging**